File: elixir/2015/lib/day4.ex
defmodule Day4 do
  @spec process(String.t(), String.t(), integer()) :: integer()
  defp process(prefix, secret, num) do
    result = :crypto.hash(:md5, "#{secret}#{num}") |> Base.encode16()

    if String.starts_with?(result, prefix) do
      num
    else
      process(prefix, secret, num + 1)
    end
  end

  @spec main() :: nil
  def main() do
    secret =
      File.read!("../../inputs/2015/Day4/input.txt") |> String.trim_trailing("\n")

    IO.inspect("Secret: #{secret}")

    result = process("00000", secret, 1)
    IO.inspect(result)

    result = process("000000", secret, 1)
    IO.inspect(result)
  end
end

File: haskell/2015/Day4.hs
module Day4 where

import Crypto.Hash.MD5
import qualified Data.Text as T
import Data.Text.Encoding
import Text.Hex

toHex :: String -> T.Text
toHex = encodeHex . hash . encodeUtf8 . T.pack

process :: T.Text -> String -> Int -> Int
process prefix s n
  | prefix `T.isPrefixOf` toHex (s ++ show n) = n
  | otherwise = process prefix s (n + 1)

main' :: IO ()
main' = do
  contents <- head . lines <$> readFile "2015/inputs/Day4/input.txt"
  -- part 1
  print $ process "00000" contents 1
  -- part 2
  print $ process "000000" contents 1
